refactor(linter): Use DefaultRuleConfig more and simplify from_configuration for tsgolint rules

connorshea opened the pull request:
This removes the `Self(Box::new())` wrapper for various tsgolint-based rules in order to simplify the implementation of the `from_configuration` method for those rules. If I understand correctly, the box wrapping is still used, it's just not referenced explicitly in the `from_configuration` method for these rules anymore, as we've moved the `DefaultRuleConfig<>` from taking the config object over to taking the rule itself.

This PR also removes various uses of `#[serde(default = "default_true")]`, as it's unnecessary when Default is defined.

Will wait to merge this until cam (either of them 😛) can review this and confirm that I'm not about to break anything here. But I'm fairly confident this should not cause any problems.

camc314 commented on crates/oxc_linter/src/rules/typescript/array_type.rs:
```diff
@@ -138,32 +138,9 @@ fn array_simple(
-                        _ => Some(ArrayOption::ArraySimple),
-                    },
-                ),
-        }))
```
yeah i think this is fine.

camc314 reviewed (changes requested):
question about default_true

connorshea committed:
refactor(linter): Use DefaultRuleConfig more and simplify from_configuration for tsgolint rules (#16818)

This removes the `Self(Box::new())` wrapper for various tsgolint-based rules in order to simplify the implementation of the `from_configuration` method for those rules. If I understand correctly, the box wrapping is still used, it's just not referenced explicitly in the `from_configuration` method for these rules anymore, as we've moved the `DefaultRuleConfig<>` from taking the config object over to taking the rule itself.

This PR also removes various uses of `#[serde(default = "default_true")]`, as it's unnecessary when Default is defined.

Will wait to merge this until cam (either of them 😛) can review this and confirm that I'm not about to break anything here. But I'm fairly confident this should not cause any problems.
